iceoryx_utils/include/iceoryx_utils/cxx/unique_ptr.hpp

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IOX_UTILS_CXX_UNIQUE_PTR_HPP
+#define IOX_UTILS_CXX_UNIQUE_PTR_HPP
+
+#include "iceoryx_utils/cxx/function_ref.hpp"
+#include <functional>
+
+namespace iox
+{
+namespace cxx
+{
+///
+/// @todo document how it differs to STL
+///
+template <typename T>
+class unique_ptr
+{
+  public:
+    using ptr_t = T*;
+
+    unique_ptr() = delete;
+
+    ///
+    /// @brief unique_ptr Creates an empty unique ptr that owns nothing. Can be passed ownership later via reset.
+    ///
+    unique_ptr(std::function<void(T*)>&& deleter) noexcept;
+
+    ///
+    /// @brief unique_ptr Creates a unique pointer that takes ownership of an object.
+    /// @details A deleter must always be provided as no default can be provided given that no head is used.
+    /// The unique_ptr must know how to delete the managed object when pointer out of scope.
+    /// @param ptr The raw pointer to the object to be managed.
+    /// @param deleter The deleter function for cleaning up the managed object.
+    ///
+    unique_ptr(ptr_t ptr, std::function<void(T*)>&& deleter) noexcept;
+
+    unique_ptr(std::nullptr_t) noexcept;
+
+    // Not copy-able to ensure uniqueness.
+    unique_ptr(const unique_ptr& other) = delete;
+    unique_ptr& operator=(const unique_ptr&) = delete;
+
+    unique_ptr(unique_ptr&& rhs) noexcept;
+    unique_ptr& operator=(unique_ptr&& rhs) noexcept;
+
+    ///
+    /// @brief operator = Reset to empty pointer when setting to nullptr.
+    /// @return An empty unique pointer.
+    ///
+    unique_ptr& operator=(std::nullptr_t) noexcept;
+
+    ///
+    /// Automatically deletes the owned object on destruction.
+    ///
+    ~unique_ptr() noexcept;
+
+    ///
+    /// Dereference the stored pointer.
+    ///
+    T operator*() noexcept;
```

Comment:
Wrong signature. You would like to return a reference here:
`T& operator*() noexcept;`

Otherwise you do the following:
```cpp
unique_ptr<int> a{new int(5)};
*a = 123;  // you get a copy instead of a reference therefore you are changing a copy of the value to 123
std::cout << *a << std::endl; // prints 5
```

Reply:
Yeah, the implementation was definitely wrong, need to think about how to do this properly - how do we handle T = void ?
Maybe also a topic for #298 ... I'm removing the dereference operator for now.